### PR TITLE
OSSM-3871 Fix Flaky test: TestCircuitBreaking

### PR DESCRIPTION
**Pull request by fjglira**
Jira: https://issues.redhat.com/browse/OSSM-3871

**TestCircuitBreaking**: change the number of requests per second to adjust and avoid condition meet of failures, no will be 10 request per second to the DestinationRule defined:

```
  trafficPolicy:
    connectionPool:
      tcp:
        maxConnections: 1
      http:
        http1MaxPendingRequests: 1
        maxRequestsPerConnection: 1
    outlierDetection:
      consecutiveErrors: 1
      interval: 1s
      baseEjectionTime: 3m
      maxEjectionPercent: 100
```

### CODE REVIEW

**Comment by luksa** [comment]
This test took 9 minutes the first time I tried it. Re-running it now.
```
--- PASS: TestCircuitBreaking (565.57s)

File: pkg/tests/tasks/traffic/circuit_breaking_test.go [hunk 1]
```diff
@@ -69,7 +69,7 @@ func TestCircuitBreaking(t *testing.T) {
 			msg := oc.Exec(t,
 				pod.MatchingSelector("app=fortio", ns),
 				"fortio",
-				fmt.Sprintf("/usr/bin/fortio load -c %d -qps 0 -n %d -loglevel Warning http://httpbin:8000/get", connection, reqCount))
+				fmt.Sprintf("/usr/bin/fortio load -c %d -qps 0.1 -n %d -loglevel Warning http://httpbin:8000/get", connection, reqCount))
```
Comment (luksa) [inline pkg/tests/tasks/traffic/circuit_breaking_test.go:72]: qps 0.1 == 1 request per 10 seconds?

Comment (fjglira) [reply pkg/tests/tasks/traffic/circuit_breaking_test.go:72]: I'm checking the documentation

Comment (fjglira) [reply pkg/tests/tasks/traffic/circuit_breaking_test.go]: I just adjust the qps to 4, increase the execution time in 10 seconds but seems to be more stable in the balance of ko vs ok

**Comment by luksa** [comment]
Looking at the docs that this test was based on (https://istio.io/latest/docs/tasks/traffic-management/circuit-breaking/), I'd say that this test needs to be changed. The number of connections should be increased to at least 3. The assertion probably shouldn't check the ratio at all, but instead just check that some requests failed, and that the istio-proxy stats show that `upstream_rq_pending_overflow` is greater than zero.

**Comment by luksa** [comment]
Also, let's add a link to the doc page, just like we've done in some other tests.

**Comment by luksa** [comment]
BTW: how is this test flaky? What's the output when it fails?

**Comment by fjglira** [comment]
We got this in failures 
```
circuit_breaking_test.go:66: STEP 5: Trip the circuit breaker by sending 50 requests to httpbin with 2 connections
    circuit_breaking_test.go:67:    We expect 60% of responses to return 200 OK, and 40% to return 503 Service Unavailable (tolerance 10%)
    circuit_breaking_test.go:82:    FAILURE: 40% of responses were 200 OK, and 60% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 1/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 36% of responses were 200 OK, and 64% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 2/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 38% of responses were 200 OK, and 62% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 3/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 26% of responses were 200 OK, and 74% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 4/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 34% of responses were 200 OK, and 66% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 5/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 28% of responses were 200 OK, and 72% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 6/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 30% of responses were 200 OK, and 70% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 7/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 26% of responses were 200 OK, and 74% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 8/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 28% of responses were 200 OK, and 72% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 9/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 40% of responses were 200 OK, and 60% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 10/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 36% of responses were 200 OK, and 64% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 11/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 30% of responses were 200 OK, and 70% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 12/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 26% of responses were 200 OK, and 74% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 13/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 32% of responses were 200 OK, and 68% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 14/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 40% of responses were 200 OK, and 60% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 15/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 36% of responses were 200 OK, and 64% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 16/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 34% of responses were 200 OK, and 66% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 17/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 36% of responses were 200 OK, and 64% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 18/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 40% of responses were 200 OK, and 60% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 19/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 38% of responses were 200 OK, and 62% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 20/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 30% of responses were 200 OK, and 70% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 21/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 44% of responses were 200 OK, and 56% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 22/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 30% of responses were 200 OK, and 70% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 23/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 36% of responses were 200 OK, and 64% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 24/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 32% of responses were 200 OK, and 68% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 25/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 26% of responses were 200 OK, and 74% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 26/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 38% of responses were 200 OK, and 62% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 27/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 30% of responses were 200 OK, and 70% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 28/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 26% of responses were 200 OK, and 74% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 29/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 30% of responses were 200 OK, and 70% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 30/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 38% of responses were 200 OK, and 62% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 31/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 30% of responses were 200 OK, and 70% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 32/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 28% of responses were 200 OK, and 72% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 33/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 36% of responses were 200 OK, and 64% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 34/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 32% of responses were 200 OK, and 68% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 35/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 28% of responses were 200 OK, and 72% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 36/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 34% of responses were 200 OK, and 66% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 37/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 30% of responses were 200 OK, and 70% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 38/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 28% of responses were 200 OK, and 72% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 39/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 30% of responses were 200 OK, and 70% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 40/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 34% of responses were 200 OK, and 66% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 41/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 34% of responses were 200 OK, and 66% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 42/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 34% of responses were 200 OK, and 66% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 43/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 28% of responses were 200 OK, and 72% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 44/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 34% of responses were 200 OK, and 66% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 45/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 30% of responses were 200 OK, and 70% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 46/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 30% of responses were 200 OK, and 70% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 47/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 30% of responses were 200 OK, and 70% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 48/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 28% of responses were 200 OK, and 72% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 49/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 26% of responses were 200 OK, and 74% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 50/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 32% of responses were 200 OK, and 68% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 51/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 34% of responses were 200 OK, and 66% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 52/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 24% of responses were 200 OK, and 76% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 53/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 28% of responses were 200 OK, and 72% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 54/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 26% of responses were 200 OK, and 74% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 55/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 36% of responses were 200 OK, and 64% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 56/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 34% of responses were 200 OK, and 66% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 57/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 38% of responses were 200 OK, and 62% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 58/60 failed. Retrying...
    circuit_breaking_test.go:82:    FAILURE: 28% of responses were 200 OK, and 72% were 503 Service Unavailable
    circuit_breaking_test.go:68:    --- Attempt 59/60 failed. Retrying...
    circuit_breaking_test.go:82:    FATAL: 26% of responses were 200 OK, and 74% were 503 Service Unavailable
```

**Comment by fjglira** [comment]
> Looking at the docs that this test was based on (https://istio.io/latest/docs/tasks/traffic-management/circuit-breaking/), I'd say that this test needs to be changed. The number of connections should be increased to at least 3. The assertion probably shouldn't check the ratio at all, but instead just check that some requests failed, and that the istio-proxy stats show that `upstream_rq_pending_overflow` is greater than zero.

Agree with you, I'll make this changes to fit the examples from Istio doc

**Comment by fjglira** [comment]
@luksa added new code now we follow the pattern like upstream documentation and check for upstream_rq_pending_overflow after made parallel request

File: pkg/tests/tasks/traffic/circuit_breaking_test.go [hunk 2]
```diff
@@ -75,12 +74,16 @@ func TestCircuitBreaking(t *testing.T) {
 			c503 := getNumberOfResponses(t, msg, `Code 503.*`)
 			successRate200 := 100 * c200 / reqCount
 			successRate503 := 100 * c503 / reqCount
+			t.Log(fmt.Sprintf("Success rate 200: %d%%", successRate200))
+			t.Log(fmt.Sprintf("Success rate 503: %d%%", successRate503))
 
-			if util.IsWithinPercentage(c200, reqCount, 0.6, tolerance) && util.IsWithinPercentage(c503, reqCount, 0.4, tolerance) {
```
Comment (luksa) [inline pkg/tests/tasks/traffic/circuit_breaking_test.go:79]: I think you should still check that at least some requests returned a 503.

File: pkg/tests/tasks/traffic/circuit_breaking_test.go [hunk 3]
```diff
@@ -75,12 +74,16 @@ func TestCircuitBreaking(t *testing.T) {
...
-			} else {
-				t.Fatalf("%d%% of responses were 200 OK, and %d%% were 503 Service Unavailable", successRate200, successRate503)
-			}
+			t.LogStep("Validate the circuit breaker is tripped by checking the istio-proxy log")
+			t.Log("Verify istio-proxy pilot-agent stats, expected upstream_rq_pending_overflow value to be more than zero")
+			output := oc.Exec(t,
+				pod.MatchingSelector("app=fortio", ns),
+				"istio-proxy",
+				"pilot-agent request GET stats | grep httpbin | grep pending")
+			assertProxyContainsUpstreamRqPendingOverflow(t, output)
```
Comment (luksa) [inline pkg/tests/tasks/traffic/circuit_breaking_test.go:86]:
```suggestion
	oc.Exec(t,
		pod.MatchingSelector("app=fortio", ns),
		"istio-proxy",
		"pilot-agent request GET stats | grep httpbin | grep pending",
		assertProxyContainsUpstreamRqPendingOverflow)
```

File: pkg/tests/tasks/traffic/circuit_breaking_test.go [hunk 4]
```diff
@@ -75,12 +74,19 @@ func TestCircuitBreaking(t *testing.T) {
...
 			successRate200 := 100 * c200 / reqCount
 			successRate503 := 100 * c503 / reqCount
-
-			if util.IsWithinPercentage(c200, reqCount, 0.6, tolerance) && util.IsWithinPercentage(c503, reqCount, 0.4, tolerance) {
-				t.LogSuccessf("%d%% of responses were 200 OK, and %d%% were 503 Service Unavailable", successRate200, successRate503)
-			} else {
-				t.Fatalf("%d%% of responses were 200 OK, and %d%% were 503 Service Unavailable", successRate200, successRate503)
+			t.Log(fmt.Sprintf("Success rate 200: %d%%", successRate200))
+			t.Log(fmt.Sprintf("Success rate 503: %d%%", successRate503))
+			if successRate503 == 0 {
```
Comment (luksa) [inline pkg/tests/tasks/traffic/circuit_breaking_test.go:79]: Is 0 ok? What if the 503 is a result of something else? 

If we expect 30% of requests to return 503, we should probably require successRate503 to be > 10%. 

Btw, why is it called "successRate503", if the request is *not* successful.

Comment (fjglira) [reply pkg/tests/tasks/traffic/circuit_breaking_test.go:79]: Ok, I'll change to validate that at least 5 request (10%) fails

**Comment by fjglira** [comment]
@luksa Added new changes

File: pkg/tests/tasks/traffic/circuit_breaking_test.go [hunk 5]
```diff
@@ -73,10 +73,10 @@ func TestCircuitBreaking(t *testing.T) {
 			c200 := getNumberOfResponses(t, msg, `Code 200.*`)
 			c503 := getNumberOfResponses(t, msg, `Code 503.*`)
 			successRate200 := 100 * c200 / reqCount
-			successRate503 := 100 * c503 / reqCount
+			failureRate503 := 100 * c503 / reqCount
 			t.Log(fmt.Sprintf("Success rate 200: %d%%", successRate200))
-			t.Log(fmt.Sprintf("Success rate 503: %d%%", successRate503))
-			if successRate503 == 0 {
+			t.Log(fmt.Sprintf("Success rate 503: %d%%", failureRate503))
+			if failureRate503 > 5 {
```
Comment (luksa) [inline pkg/tests/tasks/traffic/circuit_breaking_test.go:79]: Shouldn't this be < 5?

Comment (fjglira) [reply pkg/tests/tasks/traffic/circuit_breaking_test.go]: Done

**Comment by fjglira** [comment]
@luksa changes pushed

**Review by luksa — APPROVED** [review]

**Pull request merged**